test/integration/src/test/java/io/pravega/test/integration/ReaderGroupNotificationTest.java

```diff
@@ -265,5 +398,57 @@ public void testEndOfStreamNotifications() throws Exception {
         listenerLatch.await();
         assertTrue("Listener invoked", listenerInvoked.get());
     }
+    
+    private List<List<Segment>> scale(Controller controller, String scopeName, String streamName, int numSegments, int scalesToPerform, ScheduledExecutorService executorService) {
+        Stream stream = new StreamImpl(scopeName, streamName);
+        AtomicInteger counter = new AtomicInteger(0);
+        List<List<Segment>> listOfEpochs = new LinkedList<>();
+
+        CompletableFuture<Void> scaleFuture = Futures.loop(() -> counter.incrementAndGet() <= scalesToPerform,
+                () -> controller.getCurrentSegments(scopeName, streamName)
+                        .thenCompose(segments -> {
+                            ArrayList<Segment> sorted = Lists.newArrayList(segments.getSegments().stream()
+                                    .sorted(Comparator.comparingInt(x ->
+                                            NameUtils.getSegmentNumber(x.getSegmentId()) % numSegments))
+                                    .collect(Collectors.toList()));
+                            sorted.forEach(t -> {
+                                           log.info("SEGMENT={}", t);
+                            });
+                            listOfEpochs.add(sorted);
+                            Pair<List<Long>, Map<Double, Double>> scaleInput = getScaleInput(sorted);
+                            List<Long> segmentsToSeal = scaleInput.getKey();
+                            Map<Double, Double> newRanges = scaleInput.getValue();
+
+                            return controller.scaleStream(stream, segmentsToSeal, newRanges, executorService)
+                                    .getFuture()
+                                    .thenAccept(scaleStatus -> {
+                                        log.info("scale stream for epoch {} completed with status {}", counter.get(), scaleStatus);
+                                    });
+                        }), executorService);
+
+        scaleFuture.join();
```

Comment:
Instead of writing a future and then immediately joining it, it may be simpler to write a classing `while` loop and join on individual async futures. That's OK for writing tests.